### ISSUE TITLE
Memory leak calling map.setFeatureState() frequently [0.46-beta1]

### ISSUE THREAD

**Issue by ryanbaumann** [mapbox/mapbox-gl-js#6839]
**mapbox-gl-js version**: 0.46.0-beta1

**browser**: `Chrome v67.0`

**OS**: Windows 10

**GPU**: Geforce GTX 660 with latest driver 398.11

### Steps to Trigger Behavior

 1. Load the [example](https://bl.ocks.org/ryanbaumann/733ba99c5ca1d9d15259081b395e4b00/34d3bb174e9ac3b5e0c7f2d0cc3c41ed139849b7) in Chrome on a Windows 10 device using a Nvidia GPU (GTX GeForce 660 series with latest drivers).
 2. Move the mouse cursor repeatedly over features in the map using `map.setFeatureState()` on a vector tile source to change the hovered feature's style.  Do not zoom or pan the map.  
3. Each map.setFeatureState() call, the browser memory allocation increases.  If you call setFeatureState() enough before zooming to new tiles, the browser crashes.

### Link to Demonstration

* https://bl.ocks.org/ryanbaumann/733ba99c5ca1d9d15259081b395e4b00/34d3bb174e9ac3b5e0c7f2d0cc3c41ed139849b7

* [Video](https://www.dropbox.com/s/xn6bfdk4yx7i6pv/feature-state-crash-chrome-windows10-180620.mp4?dl=0)

* [chrome://gpu report from machine](https://gist.github.com/ryanbaumann/dd00b48287ab5aa93e6a34dc254461a0)

* [Chrome performance profile](https://www.dropbox.com/s/lcz8bwlitgqywkm/chrome-profile-mapboxgl-180620.zip?dl=0)

* ![](https://cl.ly/3F0U3a0G182P/download/Image%202018-06-20%20at%209.53.54%20AM.png)

### Expected Behavior

* Memory is reclaimed / minimally expanded for each new call of `map.setFeatureState`

### Actual Behavior

* Memory increases proportionately for each `map.setFeatureState()` call triggered
* Memory is mostly reclaimed when loading and rendering a new zoom level of tiles

cc/ @asheemmamoowala

**Comment by ChrisLoer** [comment]
@ryanbaumann Looking at that performance profile, the thing that solidly trends up is "Nodes", which I think is DOM nodes -- makes me suspect the actual problem is all the popups getting created. Can you remove the popup part of the code to see if you can still reproduce the leak?

I didn't have much luck reproducing on Chrome/Windows 10 running in VirtualBox, but I think it may just be so slow that I'm not patient enough to get the leak to show up...

**Comment by ryanbaumann** [comment]
@ChrisLoer I removed the DOM element creation on mousemove.  I and am still able to re-created the issue.

Updated example without popup DOM element - https://bl.ocks.org/ryanbaumann/733ba99c5ca1d9d15259081b395e4b00/dee9a8fa1fec66ab5bbfe856aaa6604b1e70831c

Profile: 

[Chrome file](https://www.dropbox.com/s/p05rbl8i8o2v5t0/Updated-gljs-profile.zip?dl=0)

![](https://cl.ly/280K1e2D0K3x/download/Image%202018-06-20%20at%2011.46.02%20AM.png)

GIF of memory allocation in Windows 10 for Chrome :
![screen recording 2018-06-20 at 11 41 am](https://user-images.githubusercontent.com/4342845/41678988-983e7380-7481-11e8-8d47-2587445691c0.gif)

**Comment by ChrisLoer** [comment]
Yup, looks like the popups were probably a red herring -- with that removed, the chrome profile doesn't show the "nodes" growth, but definitely looks like the same problem. Unfortunately the profile doesn't show much else going on -- JS heap usage looks like it ranges from 67-103 MB, which isn't anywhere near the several gigabyte growth for the process as a whole.

So... maybe something to do with re-uploading the the state-dependent paint vertex array to the GPU so frequently?

cc @ansis

**Comment by ChrisLoer** [comment]
This looks like the problem to me -- we re-upload when the paint arrays change, but if there's an existing vertex buffer we don't destroy it first:

https://github.com/mapbox/mapbox-gl-js/blob/20c0d13b9bb412fc18a00404212de06337350711/src/data/program_configuration.js#L186-L190

I think since these are supposed to be created as "dynamicDraw" buffers, so using `updateData` is more appropriate than destroying/recreating buffers. Updating the two `upload()`s in `program_configuration` to look more like:

```
        if (this.paintVertexArray && this.paintVertexArray.arrayBuffer) {
            if (this.paintVertexBuffer && this.paintVertexBuffer.buffer) {
                this.paintVertexBuffer.updateData(this.paintVertexArray);
            } else {
                this.paintVertexBuffer = context.createVertexBuffer(this.paintVertexArray, this.paintVertexAttributes, this.expression.isStateDependent);
            }
        }
```

Seems to have fixed the problem, at least in @ryanbaumann's test case.

**Comment by ansis** [comment]
@ChrisLoer yep! That's definitely what we need. And I'd guess this could have a pretty large performance benefit as well.

**Comment by ChrisLoer** [comment]
Fixed in #6853.
